### PR TITLE
Align colon inside ranges; Subcolumns support in tabular alignment

### LINKED ISSUE #494: Align colon inside dimension ranges
**Test case**

```systemverilog
class c;
  rand bit [DataWidth - 1   : 0]   d_data;
  rand bit [SourceWidth - 1 : 0] d_source;
endclass
```

**Actual output**

```systemverilog
class c;
  rand bit [DataWidth - 1 : 0]   d_data;
  rand bit [SourceWidth - 1 : 0] d_source;
endclass
```

Include any possible diagnostic messages from the formatter.

**Expected or suggested output**

(want the original form)

### PR DESCRIPTION
**Pull request by mglb**
Two features in a single PR to avoid dependent PRs. Each commit is one feature.

* **Add VectorTree iterators**
  Adds code that allows to traverse VectorTree in specific order using range for loop. Supported traversals:
  * Leaves
  * Pre order
  * Post order

* **Add support for subcolumns in tabular alignment mode**
  Converts columns structure from a vector to a vector tree. This allows to use subcolumns (visual example below; example use in the following commit).

* **Align dimensions and ranges in class properties and dist items**
  Fixes #494

### Demo

```
constraint num_ops_c {
  num_ops dist {
    [   1 : 1000] :/ 1,
    [1001 : 2500] :/ 6,
    [2501 : 4000] :/ 2,
    [4001 : 5000] :/ 1
  };
}
```

```
class foo extends bar;
  // C
  constraint num_trans_c {
    num_trans dist {
      baz[0]                             :/ 2,
      [       1 : 3       ]              :/ 2,
      [      14 : 25      ]              :/ 3,
      bazzzzzzzzzzzzzzzzzzzzzzzzzzz[1:4] :/ 2,
      [      25 : 37      ]              :/ 4,
      [baz[1:4] : baz[5:7]]              :/ 2,
      [      38 : 50      ]              :/ 1
    };
  }
endclass
```

```
class c;
  rand bit [ dw - 1 : 0 ]     d_data;
  rand bit [sw_ - 1 : 0 ]     d_source [             1];
  rand bit [sw_ - 1 : 0 ][ 4] d_source2[    1 : 22    ];
  rand bit                    d_source3[val+1 : oval+4];
  rand bit [sw_ - 1 : 33][22] d_source4[      CONSTANT];
endclass
```

Illustration how column tree looks for each line in the last example:
```
| rand bit |/````````````````(1+12)````````````````\|        | d_data ; |              |
|  (0+8)   |   [   |/````````(0+10)````````\|   ]   |  |  |  |  (1+7)   |  |        |  |
           | (1+1) | dw - 1 |   :   |   0   | (0+1) |                      |  |  |  |
                   | (0+6)  | (1+1) | (1+1) |


| rand bit |/````````````````(1+11)`````````````````\|        | d_source |/````````(0+4)`````````\|
|  (0+8)   |   [   |/`````````(0+9)`````````\|   ]   |  |  |  |  (1+8)   |   [   |   1    |  ] ;  |
           | (1+1) | sw_ - 1 |   :   |   0   | (0+1) |                   | (0+1) | (0+1)  | (0+2) |
                   |  (0+7)  | (0+1) | (0+1) |                                   |  |  |  |


| rand bit |/````````````````(1+11)`````````````````\|/````````(0+3)````````\| d_source2 |/````````````````(0+9)````````````````\|
|  (0+8)   |   [   |/`````````(0+9)`````````\|   ]   |   [   |   4   |   ]   |   (1+9)   |   [   |/````````(0+6)````````\|  ] ;  |
           | (1+1) | sw_ - 1 |   :   |   0   | (0+1) | (0+1) | (0+1) | (0+1) |           | (0+1) |   1   |   :   |  22   | (0+2) |
                   |  (0+7)  | (0+1) | (0+1) |                                                   | (0+1) | (1+1) | (1+2) |


| rand bit |              |        | d_source3 |/``````````````````(0+15)``````````````````\|
|  (0+8)   |  |        |  |  |  |  |   (1+9)   |   [   |/``````````(0+12)``````````\|  ] ;  |
              |  |  |  |                       | (0+1) | val + 1 |   :   | oval + 4 | (0+2) |
                                                       |  (0+5)  | (0+1) |  (0+6)   |


| rand bit |/````````````````(1+12)`````````````````\|/````````(0+4)````````\| d_source4 |/`````````(0+11)`````````\|
|  (0+8)   |   [   |/````````(0+10)`````````\|   ]   |   [   |  22   |   ]   |   (1+9)   |   [   | CONSTANT |  ] ;  |
           | (1+1) | sw_ - 1 |   :   |  33   | (0+1) | (0+1) | (0+2) | (0+1) |           | (0+1) |  (0+8)   | (0+2) |
                   |  (0+7)  | (0+1) | (0+2) |                                                   |  |   |   |
```

### CODE REVIEW

**Comment by mglb** [comment]
Added 2 small non-functional changes

File: common/formatting/align_test.cc [hunk 1]
```diff
@@ -975,5 +976,351 @@ TEST_F(InferAmbiguousAlignIntentTest, DifferenceSufficientlySmall) {
...
+    ColumnSchemaScanner::Visit(node);
+  }
+  void Visit(const SyntaxTreeLeaf& leaf) override {
+    if (!current_column_)
+      ReserveNewColumn(leaf, props);
+    else
+      ReserveNewColumn(*current_column_, leaf, props);
+  }
+
+ protected:
```
Comment (hzeller) [inline common/formatting/align_test.cc:1002]: `current_column_` being protected indicates that it is intended to be accessed in derived classes. Is this the case (I don't see any mention of current_column_ in this pull request in other classe though).

I strongly suggest to start out making everything private and once we do have to consider providing protected access, this should be a very conscious choice as it weakens the encapsulation.

**Review by hzeller — COMMENTED** [review]
Looks very nice. There are a few places where I'd use pointers instead of non-const references.

File: common/formatting/align.h [hunk 2]
```diff
@@ -66,32 +69,56 @@ struct ColumnPositionEntry {
...
+  // 'parent_column' is a reference to the parent column.
   // 'symbol' is a reference to the original source syntax subtree.
   // 'properties' contains alignment configuration for the column.
   // 'path' represents relative position within the enclosing syntax subtree,
   // and is used as a key for ordering and matching columns.
-  void ReserveNewColumn(const Symbol& symbol,
-                        const AlignmentColumnProperties& properties,
-                        const SyntaxTreePath& path);
+  ColumnPositionTree* ReserveNewColumn(
+      ColumnPositionTree& parent_column, const Symbol& symbol,
```
[the remark this answers is not in the document]
Comment (mglb) [reply common/formatting/align.h]: done

File: common/formatting/align_test.cc (hunk 1 above)
Comment (mglb) [reply common/formatting/align_test.cc]: done

File: common/formatting/align_test.cc [hunk 3]
```diff
@@ -975,5 +976,351 @@ TEST_F(InferAmbiguousAlignIntentTest, DifferenceSufficientlySmall) {
...
+      uwline.SpanUpToToken(token_iter);
+      uwline.SetOrigin(item.get());
+      partition_.NewChild(uwline);
+      SymbolCastToNode(*syntax_tree_).AppendChild(std::move(item));
+    }
+  }
+
+ private:
+  SymbolPtr ParseList(
+      std::vector<verible::PreFormatToken>::iterator& it,
```
[the remark this answers is not in the document]
Comment (mglb) [reply common/formatting/align_test.cc]: done

File: common/util/vector_tree_iterators.h [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+
+#ifndef VERIBLE_COMMON_UTIL_VECTOR_TREE_ITERATORS_H_
+#define VERIBLE_COMMON_UTIL_VECTOR_TREE_ITERATORS_H_
+
+#include <iterator>
+
+#include "common/util/iterator_range.h"
+#include "common/util/logging.h"
+#include "common/util/vector_tree.h"
+
```
Comment (hzeller) [inline common/util/vector_tree_iterators.h:23]: Maybe a little file header comment for someone scanning the file to see that this provides iterators to traverse VectorTrees in various ways ?

**Review by hzeller — CHANGES_REQUESTED** [review]
LGTM, I think mostly we should make sure to not rely too much on template argument deduction (see below), but otherwise looks very nice.

The codecov site will show which of the new lines are covered and which might need more testing; Make sure that we get good coverage, i.e. that our coverage numbers increase :)

File: common/formatting/align_test.cc [hunk 5]
```diff
@@ -975,5 +976,352 @@ TEST_F(InferAmbiguousAlignIntentTest, DifferenceSufficientlySmall) {
...
+  SyntaxTreeColumnizer() = default;
+
+  void Visit(const SyntaxTreeNode& node) override {
+    ColumnPositionTree* column;
+    if (!current_column_)
+      column = ReserveNewColumn(node, props);
+    else
+      column = ReserveNewColumn(current_column_, node, props);
+
+    ValueSaver current_column_saver(&current_column_, column);
```
[the remark this answers is not in the document]
Comment (mglb) [reply common/formatting/align_test.cc]: Done

File: common/formatting/align.cc [hunk 6]
```diff
@@ -171,53 +350,168 @@ struct AggregateColumnData {
...
+    auto parent_begin = parent_path.begin();
+    while (begin != path.end() && parent_begin != parent_path.end() &&
+           *begin == *parent_begin) {
+      ++begin;
+      ++parent_begin;
+    }
+  }
+  label << " \t ";
+  if (begin != path.begin() && begin != path.end()) label << ".";
+  label << SequenceFormatter(iterator_range(begin, path.end()), ".");
```
[the remark this answers is not in the document]
Comment (mglb) [reply common/formatting/align.cc]: Done

File: common/util/vector_tree_iterators.h (hunk 4 above)
Comment (mglb) [reply common/util/vector_tree_iterators.h:25]: Done

**Comment by mglb** [comment]
TODO: tests of new and not yet covered code

**Comment by hzeller** [comment]
Will now have another look at the recent changes.
In the meantime, the static `clang-tidy` check found a path in which it is possible to come to a division-by-zero situation - could you write a test that would exercise that path so that it is then easy to fix ?

**Comment by mglb** [comment]
>  In the meantime, the static `clang-tidy` check found a path in which it is possible to come to a division-by-zero situation - could you write a test that would exercise that path so that it is then easy to fix ?

This path is already covered by a test. I've added a CHECK_GT before this division - hope that helps. The code logic prevents this value from being zero.

**Comment by mglb** [comment]
The remaining not covered code is inside static functions that are called by `operator<<()` of non-public structs, that is used only for debug logging. How to test it? I can either move them to `align.h` (not ideal since they are implementation details) or create something like `align_internal.h`.

File: common/formatting/align_test.cc [hunk 7]
```diff
@@ -975,5 +976,450 @@ TEST_F(InferAmbiguousAlignIntentTest, DifferenceSufficientlySmall) {
...
+    }
+  }
+
+ private:
+  SymbolPtr ParseList(
+      std::vector<verible::PreFormatToken>::iterator* it,
+      const std::vector<verible::PreFormatToken>::iterator& end) {
+    SymbolPtr list = TNode(0);
+    SymbolPtr item;
+    while ((item = ParseItem(it, end)).get() != nullptr) {
```
Comment (hzeller) [inline common/formatting/align_test.cc:1063]: Isn't tha `get()` call somewhat redundant, as the smart-pointer already has an overload for comparing against `nullptr` ?

**Review by hzeller — APPROVED** [review]
LGTM.
There is a call that looks like an unnecessary get() call, please double-check.
Then ready to submit.

**Comment by hzeller** [comment]
Don't worry about not covered calls to `<<`. Goal of the coverage is to make sure all parts of the relevant algorithms are not forgotten. If a non-critical logging message here or there is not covered, that is fine.

FYI I've added some documentation and script in case you want to run coverage locally, which is arguably more convenient while developing than codecov.io
 https://github.com/chipsalliance/verible/blob/master/CONTRIBUTING.md#running-code-coverage

File: common/formatting/align_test.cc (hunk 7 above)
Comment (mglb) [reply common/formatting/align_test.cc]: It is. Removed this one and the one a few lines above (also null check)

**Pull request merged**